File: java/java_study/src/section05/Solution01.java
package section05;

import java.util.Stack;

public class Solution01 {

    public static void main(String[] args) {
        /*
        * 올바른 괄호
        *
        * 올바른 괄호이면 YES, 올바르지 않은 괄호이면 NO 출력
        * 스택 자료구조 활용 (stack)
        *
        * input : (()(()))(()
        * output : NO
        * */

        System.out.println("result : " + solution("(()(()))(()"));
    }

    public static String solution(String str) {
        String answer = "YES";
        Stack<Character> stack = new Stack<>();

        for (char c : str.toCharArray()) {
            if (c == '(') stack.push(c);
            else {
                if (stack.isEmpty()) answer = "NO";
                stack.pop();
            }
        }

        if (!stack.isEmpty()) answer = "NO";

        return answer;
    }
}

File: javascript/section06/solution01.js
/*
올바른 괄호

괄호가 입력되었을 때, 올바른 괄호이면 YES, 올바르지 않으면 NO를 출력
(())(): YES, (()())): NO

자료구조 stack 활용 - LIFO
반대개념 : queue - FIFO

input : (()(()))(()
output : NO
*/

// console.log(`result : ${solution("(()(()))(()")}`);
console.log(`result : ${solution("()()(())")}`);

function solution(str) {
    let answer = "YES";
    let stack = [];

    for(let s of str) {
        if(s === '(') stack.push(s);
        else {
            if(stack.length === 0) answer = "NO";
            stack.pop();
        }
    }

    if(stack.length > 0) answer = "NO";

    return answer;
}
